### ISSUE TITLE
Form::handleRequest integration.

### ISSUE THREAD

**Issue by duayres** [wouterj/WouterJEloquentBundle#67]
Hello,
I was trying to use internal symfony Form::handleRequest using your Eloquent bundle, but I got errors on mixins annotations inside Illuminate\Database\Eloquent\Model (not imported).
```php
        $entity = new Book(); // [Object {id: '', name: ''}]
        $form = $this->createForm(new BookType(), $entity, array(
            'action' => $this->generateUrl('book_create'),
            'method' => 'POST',
        ));

        $form->add('submit', 'submit', array('label' => 'Create'));
        
        $form->handleRequest($request);//here I get the error
```
Error: `[Semantical Error] The annotation "@mixin" in class Illuminate\Database\Eloquent\Model was never imported. Did you maybe forget to add a "use" statement for this annotation?`

Another thing is the validation, any ideas on validation using eloquent and symfony since validation on doctrine is inside the models and annotation based?

Thanks

**Comment by duayres** [comment]
This error exists due to symfony version prior to 3 mistakes the @mixin annotation.
I've partially corrected this adding after this snippet..
```
        // only process names which are not fully qualified, yet
        // fully qualified names must start with a \
        $originalName = $name;
```
...this:
```
        if ($originalName=="mixin") {
            //return as a false annotation, must not be parsed
            return false;
        }
```

But, the annotation validation (Constraints->@Assert|xxx) doesn't work, it says that the Model.name is empty everytime... :disappointed:

**Comment by wouterj** [comment]
@duayres You can add `\Doctrine\Common\Annotations\AnnotationReader::addGlobalIgnoredName('mixin');` to `Kernel::boot()` or your bootstrap file to avoid the error. We can include this in the bundle in here: https://github.com/wouterj/WouterJEloquentBundle/blob/master/src/WouterJEloquentBundle.php#L54 Are you willing to open a PR for that, so I can give you full credits?

This should btw be fixed in Doctrine annotations, so I requested a change for that: https://github.com/doctrine/annotations/pull/154 Please +1 so it has more change of getting merged :)

As for your validation issues, I've never tried that. I'll set up an example app and test it out, you'll here from me!

**Comment by duayres** [comment]
@wouterj sorry for the late answer, done the +1 on your post this morning, and now I've submitted PR #68 :) 
Thanks

**Comment by wouterj** [comment]
Late answer.. you're blazingly fast! Don't be sorry ;)

**Comment by duayres** [comment]
Ok, I've wrote a simple CRUD here and (annotation based) validation works, but I think that the form types/elements (TextType:`<input type="text>`, DateType:`<input type="date">`, etc.. ) auto-detection (symfony tries to infer the type of a input based on the data type on database) doesn't work. I think it can work if you fill your entity with `@ORM\Column(type="..")` 

PS: Ahh.. and a hint for using the annotation validation on Eloquent model entities are to put this into the config file:
```yaml
# app/config/config.yml
framework:
    validation: { enable_annotations: true }
```

**Comment by duayres** [comment]
@wouterj I've found that if i use the annotation validation setting a property on model public, the symfony validator works, but the data isn't persisted into the database by eloquent. But if you put protect or private instead, the persistence occurs but the validation fails.

After much debugging I've fount that in Validator\RecursiveContextualValidator try to extract the data from the model to validate, but it fails when it's an eloquent model.
On `RecursiveContextualValidator::validateClassNode(..)`  it receives a ClassMetadataInferface, and on line 609 we have the failing retrieval:
`$propertyValue = $propertyMetadata->getPropertyValue($object);`

propertyValue is null when its a eloquent model object.. it can't extract the value.

I've made a terrible workarround after the line that solves my problem:
```php
		if (strpos(get_class($object),"Model\\") && empty($propertyValue)) {
			$propertyValue=$object->{$propertyName};
		}
```
Here in Brazil we call it "Gambiarra" (LOL) .. a quickfix when you are tired or don't have enought time to really fix (correctly and ellegantly) a problem. Tomorrow I'll try to think in something usable that isn't a "gambiarra".

**Comment by duayres** [comment]
Update.
Another problem, but not directly related to handleRequest.
I've added a belongsTo(Author) to my Book entity:
```php
//... class Book ...
    public function author(){
        return $this->belongsTo("AppBundle\Model\Author");
    }
```
`Book::find(1)->author->name`, works perfectly on the backend, but fails when I try to add author to an FormBuilder:

```php
$builder
   ->add('nome', TextType::class, array('attr' => array('class' => 'form-control') ) )
   ->add('author');//-> it causes an error (1)
```
Adding the author field to the form triggers an internal error:
`An exception has been thrown during the rendering of a template ("Catchable Fatal Error: Object of class Illuminate\Database\Eloquent\Relations\BelongsTo could not be converted to string`

Any ideas?

**Comment by wouterj** [comment]
Hmm, so some thoughts:

 * As Eloquent requires magic properties, you can't add assertions on your model. Instead you should add the constraints to the form fields or create a [DTO object that you use for your form](https://stovepipe.systems/post/avoiding-entities-in-forms) that you can annotate with assertions.
 * You don't have to use the type guessing system with forms, so if you define all types for your properties, it should already work as expected.
 * I think form binding (form fields having the properties of the bound model object) should already work as expected.
 * We can add type guessing based on Eloquent's [attribute type casting](https://github.com/illuminate/database/blob/master/Eloquent/Concerns/HasAttributes.php#L823-L830)
 * For model relations, we probably need a custom data transformer (or even form type, just like Doctrine's EntityType) to convert from and to Eloquent Models.

**Comment by duayres** [comment]
Replying:
- In my (very humble personal) opinion, is redundance to write DTO's (you write tha same properties on database, form creation, form retrieval and validation). I was thinking of some plausible bridge between the magic properties (found in protected $attributes actual real property) and actual real POPO properties, it would be wonderful. Also i think that the exclusive behavior of eloquent is the correct way when you dont have much processing on getters and setters, you just do getters and setters for really what need to be processed (Same idea of Java Lombok). I think that my solution on that $propertyValue setting is kinda a way, but it's not elegant nor standard. Ill think of something later :smiley: 
- Again, IMHO, I think of type guessing extremely useful when you have basic CRUD's and much little time. 
- Yes, it works perfectly, thanks :grinning: 
- I'll take a look to this more deeply this weekend, I wasn't aware of this, seens very interesting.
- I'm on it... I'm basing in the propel ORM files to create the "EloquentType", "EloquentTranformer" and "EloquentChoiceLoader", right now I'm studying how it works. Lets hope my knowledge to grow and that I'll be able to create it :+1: 

Thanks

**Comment by wouterj** [comment]
@duayres I'm happy to help as well. I should have more time this weekend. I'll then also try to get Symfony 4 support in (So we can release 1.0 stable when Symfony 4 is released).

**Comment by duayres** [comment]
Just some dumb functionality I've wrote over the model I want to share:
[CustomModel Gist](https://gist.github.com/duayres/0876eafe27bfd0ea02b315d9edf3cd94) . It create a fake POPO instance of the children objects. Still missing the `toSnakeCase()` and `toCamelCase()` but it can be useful to someone... ahem.. maybe?! xD

**Comment by duayres** [comment]
@wouterj Well, I've made some progress here but I can't find a way to remap relation to relation**_id**
Here is my (UGLY code, sorry :nauseated_face:) [Gist](https://gist.github.com/duayres/5fbc54098dcf763f647ab7587cd71a0d) .

Since i'm using `builder->add('author'` (and not author**_id**), the time it will persist, author collumn can't be found on the database... so another "gambiarra" that I've mentioned earlier:

```php
    //Author (of a book) Model
    public function fixRelations(){
        foreach ($this->attributes as $attribute => $value) {
            if (substr($attribute, -3)=="_id"){
                //maps ex. 'author' to correct field 'author_id'
                $this->attributes[$attribute]=$this->attributes[substr($attribute, 0, -3)];
                unset($this->attributes[substr($attribute, 0, -3)]);
            }
        }
    }
```
Any help will be very appreciated

**Comment by wouterj** [comment]
Now I've merged https://github.com/wouterj/WouterJEloquentBundle/pull/72 , this bundle now has tested support for `Form#handleRequest()` and form type guessing. Unfortunately, I had to drop symfony 2.7 support for it, but people can seamlessly upgrade to 2.8.

Next weekend, I'll take a look at your problem with relations and eloquent form types. I would like to not mess with too much eloquent internal logic (for the simple reason that laravel internals change a lot, keeping up with it causes headages). So I'll take a look if we can manage relations with form types more easily and otherwise, I would propose to add a base form type and data transformer, which people can extend for their own use-cases.
